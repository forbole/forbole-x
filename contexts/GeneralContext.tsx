import React from 'react';
import get from 'lodash/get';
import currencies from '../misc/currencies';
import usePersistedState from '../misc/usePersistedState';

type Theme = 'light' | 'dark';
type Currency = typeof currencies[number];
type FavValidators = string[];
type FavAddress = { address: string; crypto: string; moniker: string; note?: string; img?: string };
interface GeneralState {
  currency: Currency;
  currencyRate: number; // per USD rate
  theme: Theme;
  alwaysRequirePassword: boolean;
  setAlwaysRequirePassword?: React.Dispatch<React.SetStateAction<boolean>>;
  hideAmount: boolean;
  setHideAmount?: React.Dispatch<React.SetStateAction<boolean>>;
  favValidators: FavValidators;
  favAddresses: FavAddress[];
  setCurrency?: React.Dispatch<React.SetStateAction<Currency>>;
  setTheme?: React.Dispatch<React.SetStateAction<Theme>>;
  addFavValidators?: (id: string) => void;
  deleteFavValidators?: (id: string) => void;
  addFavAddresses?: (n: FavAddress) => void;
  deleteFavAddresses?: (id: string) => void;
  updateFavAddresses?: (n: FavAddress) => void;
}

const initialState: GeneralState = {
  currency: 'USD',
  currencyRate: 1,
  theme: 'light',
  alwaysRequirePassword: false,
  hideAmount: false,
  favValidators: [],
  favAddresses: [],
};

const GeneralContext = React.createContext<GeneralState>(initialState);

const GeneralProvider: React.FC = ({ children }) => {
  const [currency, setCurrency] = usePersistedState('currency', initialState.currency);
  const [currencyRate, setCurrencyRate] = usePersistedState(
    'currencyRate',
    initialState.currencyRate,
  );
  const [theme, setTheme] = usePersistedState('theme', initialState.theme);
  const [alwaysRequirePassword, setAlwaysRequirePassword] = usePersistedState(
    'alwaysRequirePassword',
    initialState.alwaysRequirePassword,
  );
  const [hideAmount, setHideAmount] = usePersistedState('hideAmount', initialState.hideAmount);

  const [favValidators, setFavValidators] = usePersistedState('fav', initialState.favValidators);
  const [favAddresses, setFavAddresses] = usePersistedState(
    'favAddress',
    initialState.favAddresses,
  );
  const addFavValidators = React.useCallback(
    (address: string) => {
      setFavValidators(vs => [address, ...vs]);
    },
    [setFavValidators],
  );

  const deleteFavValidators = React.useCallback(
    (address: string) => {
      setFavValidators(vs => vs.filter(a => a !== address));
    },
    [setFavValidators],
  );

  const addFavAddresses = React.useCallback(
    (newAddress: FavAddress) => {
      setFavAddresses(vs => [newAddress, ...vs]);
    },
    [setFavAddresses],
  );

  const deleteFavAddresses = React.useCallback(
    (address: string) => {
      setFavAddresses(vs => vs.filter(a => a.address !== address));
    },
    [setFavAddresses],
  );

  const updateFavAddresses = React.useCallback(
    (editedAddress: {
      address: string;
      crypto: string;
      moniker: string;
      note: string;
      img: string;
      newAddress: string;
    }) => {
      setFavAddresses(vs =>
        vs.map(a =>
          a.address === editedAddress.address
            ? {
                address: editedAddress.newAddress,
                crypto: editedAddress.crypto,
                moniker: editedAddress.moniker,
                note: editedAddress.note,
                img: editedAddress.img,
              }
            : a,
        ),
      );
    },
    [setFavAddresses],
  );

  React.useEffect(() => {
    if (currency === 'USD') {
      setCurrencyRate(1);
    } else {
      fetch(
        `${process.env.NEXT_PUBLIC_ALPHAVANTAGE_API_URL}/query?function=CURRENCY_EXCHANGE_RATE&from_currency=USD&to_currency=${currency}&apikey=${process.env.NEXT_PUBLIC_ALPHAVANTAGE_API_KEY}`,
      )
        .then(r => r.json())
        .then(r =>
          setCurrencyRate(
            Number(get(r, ['Realtime Currency Exchange Rate', '5. Exchange Rate'], '')),
          ),
        );
    }
  }, [currency]);

  return (
    <GeneralContext.Provider
      value={{
        currency,
        currencyRate,
        setCurrency,
        theme,
        setTheme,
        alwaysRequirePassword,
        setAlwaysRequirePassword,
        hideAmount,
        setHideAmount,
        favValidators,
        addFavValidators,
        deleteFavValidators,
        favAddresses,
        addFavAddresses,
        deleteFavAddresses,
        updateFavAddresses,
      }}
    >
      {children}
    </GeneralContext.Provider>
  );
};

const useGeneralContext = (): GeneralState => React.useContext(GeneralContext);

export { GeneralProvider, useGeneralContext };
