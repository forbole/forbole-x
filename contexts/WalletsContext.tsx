import React from 'react';
import sendMsgToChromeExt from '../misc/sendMsgToChromeExt';
import usePersistedState from '../misc/usePersistedState';
import { useGeneralContext } from './GeneralContext';

const PASSWORD_EXPIRES_IN_MIN = 15;

interface WalletsState {
  isFirstTimeUser: boolean;
  appUnlockState: AppUnlockState;
  isChromeExtInstalled: boolean;
  wallets: Wallet[];
  accounts: Account[];
  password: string;
  updatePassword?: (password: string) => void;
  unlockWallets?: (password: string) => void;
  addWallet?: (wallet: CreateWalletParams) => void;
  updateWallet?: (id: string, wallet: UpdateWalletParams) => void;
  deleteWallet?: (id: string) => void;
  addAccount?: (account: CreateAccountParams, securityPassword: string) => void;
  updateAccount?: (address: string, walletId: string, account: UpdateAccountParams) => void;
  deleteAccount?: (address: string, walletId: string) => void;
  viewMnemonicPhrase?: (id: string, securityPassword: string) => Promise<string>;
  viewMnemonicPhraseBackup?: (
    id: string,
    securityPassword: string,
    backupPassword: string,
  ) => Promise<string>;
  reset?: () => void;
  signOut?: () => void;
}

const initialState: WalletsState = {
  isFirstTimeUser: false,
  appUnlockState: 'locked',
  isChromeExtInstalled: false,
  wallets: [],
  accounts: [],
  password: '',
};

const WalletsContext = React.createContext<WalletsState>(initialState);

const WalletsProvider: React.FC = ({ children }) => {
  const { alwaysRequirePassword } = useGeneralContext();
  const [wallets, setWallets] = React.useState<Wallet[]>([]);
  const [accounts, setAccounts] = React.useState<Account[]>([]);
  const [appUnlockState, setAppUnlockState] = React.useState(initialState.appUnlockState);
  const [isFirstTimeUser, setIsFirstTimeUser] = usePersistedState('isFirstTimeUser', false);
  const [isChromeExtInstalled, setIsChromeExtInstalled] = React.useState(false);
  const [password, setPassword] = usePersistedState(
    'password',
    '',
    PASSWORD_EXPIRES_IN_MIN * 60 * 1000,
    alwaysRequirePassword,
  );

  const reset = React.useCallback(async () => {
    await sendMsgToChromeExt({
      event: 'reset',
    });
    setIsFirstTimeUser(true);
    setAccounts([]);
    setWallets([]);
    setPassword('');
    setAppUnlockState('locked');
  }, [setIsFirstTimeUser, setAccounts, setWallets, setPassword, setAppUnlockState]);

  const checkIsFirstTimeUser = React.useCallback(async () => {
    try {
      const response = await sendMsgToChromeExt({
        event: 'ping',
      });
      setIsFirstTimeUser(response.isFirstTimeUser);
      setIsChromeExtInstalled(true);
    } catch (err) {
      setIsChromeExtInstalled(false);
    }
  }, []);

  const unlockWallets = React.useCallback(
    async (pw: string) => {
      try {
        if (!isFirstTimeUser) {
          setAppUnlockState('unlocking');
          const walletaResponse = await sendMsgToChromeExt({
            event: 'getWallets',
            data: {
              password: pw,
            },
          });
          const accountsResponse = await sendMsgToChromeExt({
            event: 'getAccounts',
            data: {
              password: pw,
            },
          });
          setWallets(walletaResponse.wallets);
          setAccounts(accountsResponse.accounts);
          setAppUnlockState('unlocked');
        }
        setPassword(pw);
      } catch (err) {
        setAppUnlockState('locked');
        throw err;
      }
    },
    [isFirstTimeUser, setPassword, setWallets, setAccounts, setAppUnlockState],
  );

  const signOut = React.useCallback(() => {
    setAppUnlockState(initialState.appUnlockState);
    setPassword(initialState.password);
    setWallets(initialState.wallets);
    setAccounts(initialState.accounts);
  }, []);

  const updatePassword = React.useCallback(
    async (pw: string) => {
      await sendMsgToChromeExt({
        event: 'changeUnlockPassword',
        data: {
          oldPassword: password,
          password: pw,
        },
      });
      setPassword(pw);
    },
    [setPassword, password],
  );

  const addWallet = React.useCallback(
    async (wallet: CreateWalletParams) => {
      const result = await sendMsgToChromeExt({
        event: 'addWallet',
        data: {
          wallet,
          password,
        },
      });
      setIsFirstTimeUser(false);
      setWallets(ws => [{ ...result.wallet, type: wallet.type }, ...ws]);
      setAccounts(acs => [...result.accounts, ...acs]);
    },
    [password, setIsFirstTimeUser, setWallets, setAccounts],
  );

  const updateWallet = React.useCallback(
    async (id: string, wallet: UpdateWalletParams) => {
      const result = await sendMsgToChromeExt({
        event: 'updateWallet',
        data: {
          wallet,
          id,
          password,
        },
      });
      setWallets(ws => ws.map(w => (w.id === id ? result.wallet : w)));
    },
    [password, setWallets],
  );

  const deleteWallet = React.useCallback(
    async (id: string) => {
      await sendMsgToChromeExt({
        event: 'deleteWallet',
        data: {
          id,
          password,
        },
      });
      await Promise.all(
        accounts
          .filter(a => a.walletId === id)
          .map(a =>
            sendMsgToChromeExt({
              event: 'deleteAccount',
              data: {
                address: a.address,
                walletId: a.walletId,
                password,
              },
            }),
          ),
      );
      setWallets(ws => {
        const newWallets = ws.filter(w => w.id !== id);
        if (!newWallets.length) {
          reset();
        }
        return newWallets;
      });
      setAccounts(acs => acs.filter(a => a.walletId !== id));
    },
    [password, accounts, setWallets, setIsFirstTimeUser, setAccounts, reset],
  );

  const addAccount = React.useCallback(
    async (account: CreateAccountParams, securityPassword: string) => {
      const result = await sendMsgToChromeExt({
        event: 'addAccount',
        data: {
          account,
          securityPassword,
          password,
        },
      });
      setAccounts(acs => [result.account, ...acs]);
    },
    [password, accounts, setAccounts],
  );

  const updateAccount = React.useCallback(
    async (address: string, walletId: string, account: UpdateAccountParams) => {
      const result = await sendMsgToChromeExt({
        event: 'updateAccount',
        data: {
          account,
          address,
          walletId,
          password,
        },
      });
      setAccounts(acs =>
        acs.map(a => (a.address === address && a.walletId === walletId ? result.account : a)),
      );
    },
    [password, setAccounts],
  );

  const deleteAccount = React.useCallback(
    async (address: string, walletId: string) => {
      await sendMsgToChromeExt({
        event: 'deleteAccount',
        data: {
          address,
          walletId,
          password,
        },
      });
      setAccounts(acs => acs.filter(a => a.address !== address || a.walletId !== walletId));
    },
    [password, setAccounts],
  );

  // Can be used to verify security password
  const viewMnemonicPhrase = React.useCallback(
    async (id: string, securityPassword: string) => {
      const { mnemonic } = await sendMsgToChromeExt({
        event: 'viewMnemonicPhrase',
        data: {
          id,
          securityPassword,
          password,
        },
      });
      return mnemonic;
    },
    [password],
  );

  const viewMnemonicPhraseBackup = React.useCallback(
    async (id: string, securityPassword: string, backupPassword: string) => {
      const { mnemonic } = await sendMsgToChromeExt({
        event: 'viewMnemonicPhraseBackup',
        data: {
          id,
          securityPassword,
          backupPassword,
          password,
        },
      });
      return mnemonic;
    },
    [password],
  );

  React.useEffect(() => {
    checkIsFirstTimeUser();
  }, []);

  return (
    <WalletsContext.Provider
      value={{
        isFirstTimeUser,
        appUnlockState,
        isChromeExtInstalled,
        wallets,
        accounts,
        password,
        unlockWallets,
        addWallet,
        updateWallet,
        deleteWallet,
        addAccount,
        updateAccount,
        deleteAccount,
        viewMnemonicPhraseBackup,
        viewMnemonicPhrase,
        reset,
        updatePassword,
        signOut,
      }}>
      {children}
    </WalletsContext.Provider>
  );
};

const useWalletsContext = (): WalletsState => React.useContext(WalletsContext);

export { WalletsProvider, useWalletsContext };
