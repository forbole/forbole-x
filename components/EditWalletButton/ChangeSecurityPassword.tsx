import {
  Button,
  DialogActions,
  DialogContent,
  DialogTitle,
  Typography,
  Box,
} from '@material-ui/core';
import useTranslation from 'next-translate/useTranslation';
import React from 'react';
import useStyles from './styles';
import { useWalletsContext } from '../../contexts/WalletsContext';
import PasswordInput from '../PasswordInput';

interface ChangeSecurityPasswordProps {
  walletId: string;
  onClose(): void;
}

const ChangeSecurityPassword: React.FC<ChangeSecurityPasswordProps> = ({ walletId, onClose }) => {
  const { t } = useTranslation('common');
  const classes = useStyles();
  const [error, setError] = React.useState('');
  const [isSettingNewPassword, setIsSettingNewPassword] = React.useState(false);
  const [securityPassword, setSecurityPassword] = React.useState('');
  const [newSecurityPassword, setNewSecurityPassword] = React.useState('');
  const { updateWallet, viewMnemonicPhrase } = useWalletsContext();

  const onButtonClick = React.useCallback(async () => {
    try {
      setError('');
      if (isSettingNewPassword) {
        await updateWallet(walletId, { newSecurityPassword, securityPassword });
        onClose();
      } else {
        await viewMnemonicPhrase(walletId, securityPassword);
        setIsSettingNewPassword(true);
      }
    } catch (err) {
      setError(err.message);
    }
  }, [
    isSettingNewPassword,
    setIsSettingNewPassword,
    securityPassword,
    newSecurityPassword,
    walletId,
    setError,
  ]);

  return (
    <form
      noValidate
      onSubmit={e => {
        e.preventDefault();
        onButtonClick();
      }}
    >
      <DialogTitle>{t('change security password')}</DialogTitle>
      <DialogContent>
        <Box mb={18}>
          <Typography gutterBottom>
            {t(isSettingNewPassword ? 'new password' : 'current password')}
          </Typography>
          <PasswordInput
            value={isSettingNewPassword ? newSecurityPassword : securityPassword}
            onChange={e =>
              (isSettingNewPassword ? setNewSecurityPassword : setSecurityPassword)(e.target.value)
            }
            placeholder={t('password')}
            error={!!error}
            helperText={error}
            withSecurityLevel={isSettingNewPassword}
          />
        </Box>
      </DialogContent>
      <DialogActions>
        <Button className={classes.dialogButton} variant="contained" color="primary" type="submit">
          {t(isSettingNewPassword ? 'save' : 'next')}
        </Button>
      </DialogActions>
    </form>
  );
};

export default ChangeSecurityPassword;
