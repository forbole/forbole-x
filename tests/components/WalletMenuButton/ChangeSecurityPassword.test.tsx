import React from 'react'
import renderer from 'react-test-renderer'
import { useWalletsContext } from '../../../contexts/WalletsContext'
import ChangeSecurityPassword from '../../../components/WalletMenuButton/ChangeSecurityPassword'

const onClose = jest.fn()
const updateWallet = jest.fn()
const viewMnemonicPhrase = jest.fn()

jest.mock('@material-ui/core/Dialog', () => (props) => <div id="dialog" {...props} />)
jest.mock('../../../contexts/WalletsContext', () => ({
  useWalletsContext: jest.fn(),
}))
jest.mock('../../../components/PasswordInput', () => 'input')
;(useWalletsContext as jest.Mock).mockReturnValue({ updateWallet, viewMnemonicPhrase })

describe('component: ChangeSecurityPasswordDialog', () => {
  it('renders opened state correctly', () => {
    const component = renderer.create(<ChangeSecurityPassword walletId="123" onClose={onClose} />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
  it('renders closed state correctly', () => {
    const component = renderer.create(<ChangeSecurityPassword walletId="123" onClose={onClose} />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
  it('calls viewMnemonicPhrase on next button click, then calls updateWallet on next button click again', async () => {
    const component = renderer.create(<ChangeSecurityPassword walletId="123" onClose={onClose} />)
    renderer.act(() => undefined)
    renderer.act(() => {
      component.root.findByType('input').props.onChange({ target: { value: 'password' } })
    })
    await renderer.act(async () => {
      await component.root.findByType('form').props.onSubmit({ preventDefault: jest.fn() })
    })
    expect(viewMnemonicPhrase).toBeCalledWith('123', 'password')
    renderer.act(() => {
      component.root.findByType('input').props.onChange({ target: { value: 'new password' } })
    })
    await renderer.act(async () => {
      await component.root.findByType('form').props.onSubmit({ preventDefault: jest.fn() })
    })
    expect(updateWallet).toBeCalledWith('123', {
      securityPassword: 'password',
      newSecurityPassword: 'new password',
    })
    expect(onClose).toBeCalled()
  })
  it('handles error on next button click', async () => {
    viewMnemonicPhrase.mockRejectedValueOnce(new Error('error'))
    const component = renderer.create(<ChangeSecurityPassword walletId="123" onClose={onClose} />)
    renderer.act(() => undefined)
    renderer.act(() => {
      component.root.findByType('input').props.onChange({ target: { value: 'password' } })
    })
    await renderer.act(async () => {
      await component.root.findByType('form').props.onSubmit({ preventDefault: jest.fn() })
    })
    expect(viewMnemonicPhrase).toBeCalledWith('123', 'password')
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})

afterEach(() => {
  jest.clearAllMocks()
})
