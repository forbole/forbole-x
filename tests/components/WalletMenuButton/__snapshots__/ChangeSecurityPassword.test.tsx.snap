// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component: ChangeSecurityPasswordDialog handles error on next button click 1`] = `
<form
  noValidate={true}
  onSubmit={[Function]}
>
  <div
    className="MuiDialogTitle-root"
  >
    <h2
      className="MuiTypography-root MuiTypography-h6"
    >
      common:change security password
    </h2>
  </div>
  <div
    className="MuiDialogContent-root"
  >
    <div
      className="MuiBox-root MuiBox-root-18"
    >
      <p
        className="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom"
      >
        common:current password
      </p>
      <input
        error={true}
        helperText="error"
        onChange={[Function]}
        placeholder="common:password"
        value="password"
        withSecurityLevel={false}
      />
    </div>
  </div>
  <div
    className="MuiDialogActions-root MuiDialogActions-spacing"
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained HookGlobalStyles-dialogButton-4 MuiButton-containedPrimary"
      disabled={false}
      onBlur={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="submit"
    >
      <span
        className="MuiButton-label"
      >
        common:next
      </span>
      <span
        className="MuiTouchRipple-root"
      />
    </button>
  </div>
</form>
`;

exports[`component: ChangeSecurityPasswordDialog renders closed state correctly 1`] = `
<form
  noValidate={true}
  onSubmit={[Function]}
>
  <div
    className="MuiDialogTitle-root"
  >
    <h2
      className="MuiTypography-root MuiTypography-h6"
    >
      common:change security password
    </h2>
  </div>
  <div
    className="MuiDialogContent-root"
  >
    <div
      className="MuiBox-root MuiBox-root-16"
    >
      <p
        className="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom"
      >
        common:current password
      </p>
      <input
        error={false}
        helperText=""
        onChange={[Function]}
        placeholder="common:password"
        value=""
        withSecurityLevel={false}
      />
    </div>
  </div>
  <div
    className="MuiDialogActions-root MuiDialogActions-spacing"
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained HookGlobalStyles-dialogButton-4 MuiButton-containedPrimary"
      disabled={false}
      onBlur={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="submit"
    >
      <span
        className="MuiButton-label"
      >
        common:next
      </span>
      <span
        className="MuiTouchRipple-root"
      />
    </button>
  </div>
</form>
`;

exports[`component: ChangeSecurityPasswordDialog renders opened state correctly 1`] = `
<form
  noValidate={true}
  onSubmit={[Function]}
>
  <div
    className="MuiDialogTitle-root"
  >
    <h2
      className="MuiTypography-root MuiTypography-h6"
    >
      common:change security password
    </h2>
  </div>
  <div
    className="MuiDialogContent-root"
  >
    <div
      className="MuiBox-root MuiBox-root-15"
    >
      <p
        className="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom"
      >
        common:current password
      </p>
      <input
        error={false}
        helperText=""
        onChange={[Function]}
        placeholder="common:password"
        value=""
        withSecurityLevel={false}
      />
    </div>
  </div>
  <div
    className="MuiDialogActions-root MuiDialogActions-spacing"
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained HookGlobalStyles-dialogButton-4 MuiButton-containedPrimary"
      disabled={false}
      onBlur={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="submit"
    >
      <span
        className="MuiButton-label"
      >
        common:next
      </span>
    </button>
  </div>
</form>
`;
